protobuf/plugin/deepcompare/deepcompare.go

```diff
@@ -0,0 +1,293 @@
+	d.Out()
+	d.P("}")
+
+	oneofByIndex := [][]*descriptor.FieldDescriptorProto{}
+	for _, f := range m.Field {
+		fName := generator.CamelCase(*f.Name)
+		if gogoproto.IsCustomName(f) {
+			fName = gogoproto.GetCustomName(f)
+		}
+
+		// Handle oneof type, we defer them to a loop below
+		if f.OneofIndex != nil {
+			if len(oneofByIndex) <= int(*f.OneofIndex) {
+				oneofByIndex = append(oneofByIndex, []*descriptor.FieldDescriptorProto{})
+			}
+
+			oneofByIndex[*f.OneofIndex] = append(oneofByIndex[*f.OneofIndex], f)
+			continue
+		}
+
+		// Handle all kinds of message type
+		if f.IsMessage() {
+			// Handle map type
+			if d.genMap(m, f) {
+				continue
+			}
+
+			// Handle any message which is not repeated or part of oneof
+			if !f.IsRepeated() && f.OneofIndex == nil {
+				if !gogoproto.IsNullable(f) {
+					d.genEqualFunc("&m."+fName, "&o."+fName)
+				} else {
+					d.genEqualFunc("m."+fName, "o."+fName)
+				}
+				continue
+			}
+		}
+
+		// Handle repeated field
+		if f.IsRepeated() {
+			d.genRepeated(m, f)
+			continue
+		}
+
+		// Handle bytes
+		if f.IsBytes() {
+			d.genBytesEqual("m."+fName, "o."+fName)
+			continue
+		}
+
+		// scalar: normal equality
+		d.genScalarEqual("m."+fName, "o."+fName)
+	}
+
+	for i, oo := range m.GetOneofDecl() {
+		d.genOneOf(m, oo, oneofByIndex[i])
+	}
+
+	d.P("return true")
+	d.Out()
+	d.P("}")
+	d.P()
+}
+
+func (d *deepCompareGen) genMap(m *generator.Descriptor, f *descriptor.FieldDescriptorProto) bool {
+	fName := generator.CamelCase(*f.Name)
+	if gogoproto.IsCustomName(f) {
+		fName = gogoproto.GetCustomName(f)
+	}
+
+	dv := d.ObjectNamed(f.GetTypeName())
+	desc, ok := dv.(*generator.Descriptor)
+	if !ok || !desc.GetOptions().GetMapEntry() {
+		return false
+	}
+
+	mt := d.GoMapType(desc, f)
+
+	d.genScalarEqual("len(m."+fName+")", "len(o."+fName+")")
+	d.P("for k, vo := range o.", fName, " {")
+	d.In()
+	d.P("vm, ok := m.", fName, "[k]")
+	d.P("if !ok {")
+	d.In()
+	d.P("return false")
+	d.Out()
+	d.P("}")
+	if mt.ValueField.IsMessage() {
+		if !gogoproto.IsNullable(f) {
+			d.genEqualFunc("&vo", "&vm")
+		} else {
+			d.genEqualFunc("vo", "vm")
+		}
+	} else if mt.ValueField.IsBytes() {
+		d.genBytesEqual("vo", "vm")
+	} else {
+		d.genScalarEqual("vo", "vm")
+	}
+	d.Out()
+	d.P("}")
+	d.P()
+
+	return true
+}
+
+func (d *deepCompareGen) genRepeated(m *generator.Descriptor, f *descriptor.FieldDescriptorProto) {
+	fName := generator.CamelCase(*f.Name)
+	if gogoproto.IsCustomName(f) {
+		fName = gogoproto.GetCustomName(f)
+	}
+
+	d.genScalarEqual("len(m."+fName+")", "len(o."+fName+")")
+
+	if proto.GetBoolExtension(f.Options, plugin.E_OrderInsensitive, false) {
+		// Order insensitive compare
+		d.P("remaining", fName, " := make(map[int]struct{})")
+		d.P("for i := range o.", fName, " {")
+		d.In()
+		d.P("remaining", fName, "[i] = struct{}{}")
+		d.Out()
+		d.P("}")
+		d.P("outer", fName, " :")
+		d.P("for i := range m.", fName, " {")
+		d.In()
+		d.P("for j := range remaining", fName, " {")
```

Comment:
I couldn't think of a way to do this. We can't index the objects with a map directly because they may be messages that contain pointers, slices, etc. Let me know if you have any ideas.

The best idea I have right now is to convert the `Equal` method to a sort comparator, and then do an out-of-place sort. But I don't think it's worth it right now, since all the relevant repeated fields should have just a few items.